---
title: "RNA-seq analysis - CHIKV infection"
editor_options: 
  chunk_output_type: console
---

```{r Load libraries}
library(here)
library(tidyverse)
library(edgeR)
library(EDASeq)
library(pheatmap)
library(ggfortify)
library(RColorBrewer)
library(fgsea)
library(patchwork)
library(ggrepel)

options(scipen=999)
```

```{r Load files}
# Immune response genes
immgene <- read_tsv(here("data", "immune_response_GO0006955_genes.txt"),
                    col_names = F)

# Human genome annotation
hannot <- read_tsv(here("data", 
                        "Homo_sapiens.GRCh38.100_gene_annotation_table.txt"))

# Read phenodata
pheno <- read_tsv(here("data", 'phenodata', "phenodata_revision.tsv")) %>% 
  mutate(Class = factor(Class), Day  = factor(Day))

# Read gene count matrix
rawcount <- read_tsv(here("data", "geneCounts",
                          "rawCounts_featureCounts_NOTmultimapping.tsv")) %>%
  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
              ends_with("bam")) %>% 
  rename_with(~ sub("19017", "", .x), everything()) %>% 
  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>% 
  dplyr::select("Symbol", as.character(pheno$Sample))
  
#write_tsv(rawcount, here("data","geneCounts",
#               "rawCounts_featureCounts_NOTmultimapping_cleanNames.txt"),
#          col_names = F)
```

```{r TMM normalization}
group <- factor(paste(pheno$Class, pheno$Day, sep = "_"))

expr <- DGEList(counts = rawcount[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

expr.norm.fact <- calcNormFactors(expr, method = 'TMM')

tmm.counts <- cpm(expr.norm.fact, log = F)
out.tmm.counts <- cbind(hannot[,c(1,2,4)], tmm.counts)

rownames(tmm.counts) <- expr.norm.fact$genes$gene_id
#write_tsv(out.tmm.counts,
#          here("data", "gene_expression", '98libs',
#               "counts_NOTmultimapping_TMM_allGenes.tsv"))
```

```{r Heatmap}
V <- apply(tmm.counts, 1, var)

# Select top 500 most variant genes
selectedGenes <- names(V[order(V, decreasing = T)][1:500])

annot <- pheno[,2:4]
rownames(annot) <- pheno$Sample

colors <- colorRampPalette(c("blue","white","red"))

heatmap <- pheatmap(log1p(tmm.counts[selectedGenes,]), scale = 'row', 
                    show_rownames = F, annotation_col = as.data.frame(annot),
                    show_colnames = F, 
                    color = colors(100))

#ggsave(here("results","figures", '98libs', "Heatmap_top-variance-genes.png"), heatmap, 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 2)
```

```{r RLE}
colors <- brewer.pal(3, "Set2")

# Not multimapping
plotRLE(as.matrix(rawcount[,-1]), outline = FALSE, 
        col = colors[pheno$Class], ylim = c(-4, 4), 
        main = "Raw counts", 
        ylab= "Relative Log Expression")

plotRLE(tmm.counts, outline=FALSE, col = colors[pheno$Class],
        ylim = c(-2, 2), main = "TMM", 
        ylab= "Relative Log Expression")
```

```{r PCA}
V <- apply(tmm.counts, 1, var)

selectedGenes <- names(V[order(V, decreasing = T)][1:500])

annot <- pheno[,2:4]
rownames(annot) <- pheno$Sample

# Not multimapping CPM
M <- t(tmm.counts[selectedGenes,])

M <- log1p(M)

pcaResults <- prcomp(M, scale. = T)

pca <- autoplot(pcaResults, data = as.data.frame(annot), 
         colour = 'Day', shape = "Class") +
  scale_color_manual(values = brewer.pal(3, "Dark2")[2:3]) +
  theme_bw()

#ggsave(here("results","figures",'98libs', "PCA_TMM.png"), pca, 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 1)
```

```{r DE analysis - edgeR}
group <- factor(paste(pheno$Class, pheno$Day, sep = "_"))

y <- DGEList(counts = rawcount[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

# Design matrix
design <- model.matrix(~ 0 + group)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 1)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)
#plotMDS(y, col = brewer.pal(4,"Set2")[group], labels = group)

tmm <- cpm(y)
rownames(tmm) <- y$genes$gene_id

out.tmm <- tmm %>% 
  as.data.frame() %>% 
  rownames_to_column('gene_id') %>% 
  left_join(hannot[,c(1,2,4)]) %>% 
  relocate("gene_id", "GeneSymbol", "Class")

out.tmm.td <- out.tmm %>% 
  pivot_longer(-c('gene_id', 'GeneSymbol', 'Class'), names_to = 'Sample') %>% 
  left_join(pheno %>% mutate(Sample = as.character(Sample)), by = 'Sample')

#write_tsv(out.tmm,
#          here("data", "gene_expression", '98libs', 
#               "counts_NOTmultimapping_TMM_filtered_20k.tsv"))

# Estimate Dispersion
y <- estimateDisp(y, design)
#plotBCV(y)

#DE
fit <- glmQLFit(y, design)

# Contrast matrix
my.contrast <- makeContrasts(
  chronVSnonChron.D0 = groupChronic_D0 - groupnonChronic_D0,
  chronVSnonChron.D21 = groupChronic_D21 - groupnonChronic_D21,
  chronD0VSchronD21 = groupChronic_D0 - groupChronic_D21,
  nonChronD0VSnonChronD21 = groupnonChronic_D0 - groupnonChronic_D21,
  levels = design)

chronVSnonChron.D0 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSnonChron.D0"])
chronVSnonChron.D21 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSnonChron.D21"])
chronD0VSchronD21 <- glmQLFTest(fit, contrast = my.contrast[, "chronD0VSchronD21"])
nonChronD0VSnonChronD21 <- glmQLFTest(fit, 
                                      contrast = my.contrast[, "nonChronD0VSnonChronD21"])

chronVSnonChron.D0_t <- topTags(chronVSnonChron.D0, n = NULL)$table
chronVSnonChron.D21_t <- topTags(chronVSnonChron.D21, n = NULL)$table
chronD0VSchronD21_t <- topTags(chronD0VSchronD21, n = NULL)$table
nonChronD0VSnonChronD21_t <- topTags(nonChronD0VSnonChronD21, n = NULL)$table


meanTMM <- cbind(y$genes[,1:3],
                 rowMeans(tmm[,group == 'nonChronic_D0']), 
                 rowMeans(tmm[,group == 'nonChronic_D21']),
                 rowMeans(tmm[,group == 'Chronic_D0']),
                 rowMeans(tmm[,group == 'Chronic_D21']))

colnames(meanTMM) <- c("gene_id", "GeneSymbol", "Class",
                       'nonChronic_D0_TMM', 'nonChronic_D21_TMM', 
                       'Chronic_D0_TMM', 'Chronic_D21_TMM')

#write_tsv(meanTMM,
#          here("results", "DEG_edgeR", '98libs',
#               "meanTMM_by_group.tsv"))


########## Join
edger.output <- meanTMM %>% 
  full_join(chronVSnonChron.D0_t, by = c("gene_id","GeneSymbol","Class")) %>% 
  full_join(chronVSnonChron.D21_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronVSnonChron.D0", "_chronVSnonChron.D21")) %>% 
  full_join(chronD0VSchronD21_t, by = c("gene_id","GeneSymbol","Class")) %>% 
  full_join(nonChronD0VSnonChronD21_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronD0VSchronD21", "_nonChronD0VSnonChronD21"))

edgeR.out <- bind_rows("chronVSnonChron.D0" = chronVSnonChron.D0_t,
                       "chronVSnonChron.D21" = chronVSnonChron.D21_t,
                       "chronD0VSchronD21" = chronD0VSchronD21_t,
                       "nonChronD0VSnonChronD21" = nonChronD0VSnonChronD21_t,
                      .id = "compar")

#write_tsv(edgeR.out, here("results", "DEG_edgeR", '98libs',
#                          "arbobios_phase_condition.tsv"))

#write_tsv(edger.output,
#          here("results", "DEG_edgeR", '98libs',
#               "diff_express_all_comparisons.tsv"))

edger.output %>% 
  filter(logFC_chronVSnonChron.D0 > 1 & FDR_chronVSnonChron.D0 < 0.1) %>% nrow()
edger.output %>% 
  filter(logFC_chronVSnonChron.D0 < -1 & FDR_chronVSnonChron.D0 < 0.1) %>% nrow()

edger.output %>% 
  filter(logFC_chronVSnonChron.D21 > 1 & FDR_chronVSnonChron.D21 < 0.1) %>% nrow()
edger.output %>% 
  filter(logFC_chronVSnonChron.D21 < -1 & FDR_chronVSnonChron.D21 < 0.1) %>% nrow()

edger.output %>% 
  filter(logFC_chronD0VSchronD21 > 1 & FDR_chronD0VSchronD21 < 0.1) %>% nrow()
edger.output %>% 
  filter(logFC_chronD0VSchronD21 < -1 & FDR_chronD0VSchronD21 < 0.1) %>% nrow()

edger.output %>% 
  filter(logFC_nonChronD0VSnonChronD21 > 1 & 
           FDR_nonChronD0VSnonChronD21 < 0.1) %>% nrow()
edger.output %>% 
  filter(logFC_nonChronD0VSnonChronD21 < -1 & 
           FDR_nonChronD0VSnonChronD21 < 0.1) %>% nrow()
```

```{r Volcano Plot}
##### Volcano plot function (NO label)
volc_plot <- function(topTags.out, df = immgene) {
  up <- topTags.out %>% filter(logFC > 1 & FDR < 0.1)
  
  down <- topTags.out %>% filter(logFC < -1 & FDR < 0.1)
  
  topTags.out %>% 
    ggplot(aes(logFC, -log10(FDR))) +
    geom_point(color = "grey", size = 0.8) +
    geom_point(data = up ,
               aes(logFC, -log10(FDR)), color = "#EF8A62", size = 0.8) +
    geom_point(data = down,
               aes(logFC, -log10(FDR)), color = "#67A9CF", size = 0.8) +
    geom_vline(xintercept = c(-log2(2), log2(2)), linetype="dashed", 
                  color = "red", size = 0.3) +
    geom_hline(yintercept = -log10(0.1), linetype="dashed", color = "red", 
               size = 0.3) +
    theme_classic()
}

# chronVSctrl.D0
chronVSnonChron.D0_vp <- volc_plot(chronVSnonChron.D0_t)

# chronVSctrl.D21
chronVSnonChron.D21_vp <- volc_plot(chronVSnonChron.D21_t)

# chronD21VSchronD0
chronD0VSchronD21_vp <- volc_plot(chronD0VSchronD21_t)

# ctrlD21VSctrlD0
nonChronD0VSnonChronD21_vp <- volc_plot(nonChronD0VSnonChronD21_t)


patchw <- (chronVSnonChron.D0_vp | chronVSnonChron.D21_vp)/
  (chronD0VSchronD21_vp | nonChronD0VSnonChronD21_vp)

fp <- patchw + plot_annotation(tag_levels = 'A')

#ggsave(here("results","figures", '98libs', "volcanoPlot_all_comparisons.png"), fp, 
#       device = "png", width = 9, height = 8, units = "cm", 
#       dpi = "print", scale = 2.5)

###### Volcano plot function WITH label
volc_plot_la <- function(topTags.out, df = immgene) {
  up <- topTags.out %>% filter(logFC > 1 & FDR < 0.1)
  
  down <- topTags.out %>% filter(logFC < -1 & FDR < 0.1)
  
  topTags.out %>% 
    ggplot(aes(logFC, -log10(FDR))) +
    geom_point(color = "grey") +
    geom_point(data = up ,
               aes(logFC, -log10(FDR)), color = "#EF8A62", size = 2.5) +
    geom_point(data = down,
               aes(logFC, -log10(FDR)), color = "#67A9CF", size = 2.5) +
    geom_vline(xintercept = c(-log2(2), log2(2)), linetype="dashed", 
                  color = "red", size = 0.3) +
    geom_hline(yintercept = -log10(0.1), linetype="dashed", color = "red", 
               size = 0.3) +
    geom_text_repel(data = up %>%
                      bind_rows(down) %>% 
                      filter(GeneSymbol %in% df$X1), 
                    aes(label = GeneSymbol), min.segment.length = 0,
                    max.overlaps = Inf, size = 3) +
    theme_classic()
}

# chronVSctrl.D0
chronVSnonChron.D0_vp_la <- volc_plot_la(chronVSnonChron.D0_t)

#ggsave(here("results","figures", '98libs', "volcanoPlot_chronVSnonChron.D0.png"),
#       chronVSnonChron.D0_vp_la , 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 2.5)

# chronVSctrl.D21
chronVSnonChron.D21_vp_la <- volc_plot_la(chronVSnonChron.D21_t)

#ggsave(here("results","figures",'98libs',"volcanoPlot_chronVSnonChron.D21.png"),
#       chronVSnonChron.D21_vp_la, 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 2.5)

# chronD21VSchronD0
chronD0VSchronD21_vp_la <- volc_plot_la(chronD0VSchronD21_t)

#ggsave(here("results","figures",'98libs',"volcanoPlot_chronD0VSchronD21.png"),
#       chronD0VSchronD21_vp_la , 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 2.5)

# ctrlD21VSctrlD0
nonChronD0VSnonChronD21_vp_la <- volc_plot_la(nonChronD0VSnonChronD21_t)

#ggsave(here("results","figures",'98libs',"volcanoPlot_nonChronD0VSnonChronD21.png"),
#       nonChronD0VSnonChronD21_vp_la, 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 2.5)

```

```{r FGSEA Reactome}
gene_set <- gmtPathways(here("data", "reactome_data", 
                             "ReactomePathwaysLevel3_2021-03.gmt"))

root_node <- read_tsv(here("data", "reactome_data", 
                           "ReactomePathways_leave_root_third_level.tsv"))

# chronVSnonChron.D0
ranks_chronVSnonChron.D0 <- chronVSnonChron.D0_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(logFC=mean(rank)) %>% 
  deframe()

fgseaRes_chronVSnonChron.D0 <- fgsea(pathways=gene_set,
                                       stats=ranks_chronVSnonChron.D0, eps=0)

# chronVSnonChron.D21
ranks_chronVSnonChron.D21 <- chronVSnonChron.D21_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chronVSnonChron.D21 <- fgsea(pathways=gene_set, 
                                       stats=ranks_chronVSnonChron.D21, eps=0)


# chronD0VSchronD21
ranks_chronD0VSchronD21 <- chronD0VSchronD21_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chronD0VSchronD21 <- fgsea(pathways=gene_set, 
                                    stats=sort(ranks_chronD0VSchronD21), eps=0)

# nonChronD0VSnonChronD21
ranks_nonChronD0VSnonChronD21 <- nonChronD0VSnonChronD21_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_nonChronD0VSnonChronD21 <- fgsea(pathways=gene_set, 
                                    stats=ranks_nonChronD0VSnonChronD21, eps=0)


# Join fgsea results
fgseaRes <- bind_rows("chronVSnonChron.D0" = fgseaRes_chronVSnonChron.D0, 
                      "chronVSnonChron.D21" = fgseaRes_chronVSnonChron.D21,
                      "chronD0VSchronD21" = fgseaRes_chronD0VSchronD21,
                      "nonChronD0VSnonChronD21" = fgseaRes_nonChronD0VSnonChronD21,
                      .id = "compar") %>% 
  left_join(root_node %>% 
              dplyr::select(root_name, node_name), 
            by = c("pathway" = "node_name")) %>% 
  relocate(compar, pathway, root_name)

custom_pal <- colorRampPalette(c('blue', 'white', 'red'))(200)

# chronic vs nonChronic
fgseaRes_phase <- fgseaRes %>% 
  filter(compar %in% c("chronVSnonChron.D0","chronVSnonChron.D21")) %>% 
  group_by(pathway) %>% 
  filter(count(padj < 0.01) == 2) %>% 
  mutate(NES = ifelse(padj > 0.01, 0, NES)) %>% 
  na.omit()

max_abs_nes <- max(abs(fgseaRes_phase$NES))

fgseaRes_phase %>% 
  ggplot(aes(compar, y=pathway, size=-log10(padj), fill=NES)) +
  geom_point(color = 'white', shape=21) +
  scale_fill_gradientn(colours=custom_pal, space = "Lab",
                       limits=c(-max_abs_nes, max_abs_nes),
                       ) +
  facet_grid(rows = vars(root_name), scales = "free_y", space = "free_y",
             switch = "y") +
  guides(size = guide_legend(override.aes = list(fill = "black")))+
  theme_bw() +
  theme(panel.grid.major = element_line(linetype = "dashed"),
        panel.spacing = unit(0, "lines"),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust=1),
        strip.text.y.left = element_text(angle = 0),
        strip.background =element_rect(fill=NA))

# Both
fgseaRes_all <-  fgseaRes %>% 
  group_by(pathway) %>% 
  filter(count(padj < 1e-5) >= 3) %>% 
  mutate(NES = ifelse(padj > 1e-5, 0, NES)) %>% 
  na.omit() %>% 
  mutate(compar = factor(compar, 
                            level = c( "chronVSnonChron.D0",
                                       "chronVSnonChron.D21",
                                       "chronD0VSchronD21", 
                                       "nonChronD0VSnonChronD21")))

max_abs_nes <- max(abs(fgseaRes_all$NES))

arbo_phase_chron_gsea <- fgseaRes_all %>% 
  ggplot(aes(compar, y=pathway, size=-log10(padj), fill=NES)) +
  geom_point(color = 'white', shape=21) +
  scale_fill_gradientn(colours=custom_pal, space = "Lab",
                       limits=c(-max_abs_nes, max_abs_nes),
                       ) +
  facet_grid(rows = vars(root_name), scales = "free_y", space = "free_y",
             switch = "y") +
  guides(size = guide_legend(override.aes = list(fill = "black")))+
  theme_bw() +
  theme(panel.grid.major = element_line(linetype = "dashed"),
        panel.spacing = unit(0, "lines"),
        axis.text.x = element_text(angle = 35, vjust = 1, hjust=1),
        strip.text.y.left = element_text(angle = 0),
        strip.background =element_rect(fill=NA)) +
  scale_x_discrete(name = NULL,
                   labels= c( "Chron VS non-Chron - Early",
                             "Chron VS non-Chron - Late",
                             "Chron.Early VS Chron.Late",
                             "non-Chron.Early VS non-Chron.Late")) +
  scale_y_discrete(name = NULL)

#ggsave(here("results","figures","98libs", "arbo_phase_chronicity_gsea.png"),
#       arbo_phase_chron_gsea,
#       device = "png", 
#       dpi = "print", scale = 1.4)
```
