---
title: "Data integration"
editor_options: 
  chunk_output_type: console
---

```{r Load libraries}
library(here)
library(tidyverse)
library(edgeR)
library(ggfortify)
library(RColorBrewer)
library(sva)
library(CEMiTool)
library(viridis)
```

```{r Load files}
# CSBL CHIKV phenodata
sra.csbl <- read_csv(here("data", "chikv_csbl_star", "SraRunTable.txt"))

sup.csbl <- read_tsv(here("data", "chikv_csbl_star", "ppat.1007880.s001.csv"))

# Arbobios phenodata
pheno.arbo <- read_tsv(here("data", "phenodata.tsv"))

# Human genome annotation
hannot <- read_tsv(here("data", 
                        "Homo_sapiens.GRCh38.100_gene_annotation_table.txt"))

# CSBL read count
counts_csbl <- read_tsv(here("data", "chikv_csbl_star", "CountTable.tsv")) %>% 
  rename(Symbol = genes)

# Arbobobios read count
counts_arbo <- read_tsv(here("data", "geneCounts",
                          "rawCounts_featureCounts_NOTmultimapping.tsv")) %>%
  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
              ends_with("bam")) %>% 
  rename_with(~ sub("19017", "", .x), everything()) %>% 
  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>%
  rename_with(~ paste0("ARB", .x), -Symbol) %>%
  dplyr::select("Symbol", sort(colnames(.)))

gmt.reactn <- read_gmt(here("data", "reactome_data", 
                            "ReactomePathwaysLevel3_2021-03.gmt"))

int_df <- read_tsv(here("data", "interaction_DB", 
                        "9606.gene.physical.links.STRINGv11_HGNC.tsv")) %>% 
  as.data.frame()

#adj.raw.counts <- read_tsv(here("data", "gene_expression", 
#                                "Arbobios_CSBL_rawCounts_CombatAdjusted.tsv"))

```

```{r Join CSBL and Arbobios phenodata + counts}
pheno.csbl <- sra.csbl %>% 
  full_join(sup.csbl, by = c("Sample Name" = "Subject_ID")) %>%
  mutate(condition = if_else(Sample_type == "Control", "healthy", 
                             "infected")) %>% 
  dplyr::select(Run, Label, Days_of_symptoms_Cat, Arthralgia_Chronicity,
                condition) %>% 
  mutate(day = if_else(Days_of_symptoms_Cat == "Control", NA_character_, 
                        Days_of_symptoms_Cat)) %>% 
  mutate(disease.status = case_when(Arthralgia_Chronicity == "ND" ~ NA_character_,
                             Arthralgia_Chronicity == "Y" ~ "Chronic",
                             Arthralgia_Chronicity == "N" ~ "nonChronic"))

pheno <- pheno.arbo %>%
  rename(Run = Sample, day = Day) %>% 
  mutate(disease.status = ifelse(Class == "control", "nonChronic", Class)) %>% 
  mutate(condition = "infected") %>% 
  dplyr::select(-Class, - Batch) %>%
  mutate(Run = paste0("ARB", Run)) %>% 
  bind_rows(pheno.csbl %>% dplyr::select(Run, condition, day, 
                                         disease.status)) %>%
  mutate(phase = case_when(day == "D21" ~ "late",
                           day == "D20" ~ "late",
                           day == "D0" ~ "early",
                           day == "D0or1" ~ "early",
                           day == "D2" ~ "early",
                           day == "D3or4" ~ "early",
                           TRUE ~ NA_character_)) %>% 
  relocate(Run, condition, day, phase, disease.status) %>% 
  mutate(dataset = case_when(day == "D0" ~ "arbobios",
                           day == "D21" ~ "arbobios",
                           TRUE ~ "csbl")) %>% 
  mutate(condition = as.factor(condition)) %>% 
  mutate(dataset = as.factor(dataset))

#write_tsv(pheno, here("data", "Arbobios_CSBL_CHIKV_phenodata.tsv"))

rawcount <- counts_arbo %>% 
  full_join(counts_csbl, by = "Symbol") %>% 
  relocate(Symbol, pheno$Run)
```

```{r Normalization + PCA}
expr <- DGEList(counts = rawcount[,-1], 
                genes = hannot[,c(1,2,4)],
                group = paste(pheno$condition, pheno$phase, 
                              pheno$disease.status, sep = "_"))

expr.norm <- calcNormFactors(expr, method = "TMM")

tmm.counts <- cpm(expr.norm, log = F)
cpm.counts <- cpm(expr, log = F)

rownames(tmm.counts) <- rawcount$Symbol
rownames(cpm.counts) <- rawcount$Symbol

annot <- pheno %>% 
  dplyr::select(Run, condition, phase, dataset, disease.status) %>% 
  mutate(phase = ifelse(is.na(phase) == T, "healthy", phase)) %>% 
  mutate(disease.status = case_when(phase == "healthy" ~ "healthy", 
                                    phase != 'healthy' &
                                    is.na(disease.status) ~ 'unknown',
                                    TRUE ~ disease.status)) %>% 
  column_to_rownames('Run')

# Select most variant genes
genes.var <- apply(tmm.counts, 1, var)
selectedGenes <- names(genes.var[order(genes.var, decreasing = T)][1:500])

M <- log2(t(tmm.counts[selectedGenes,]) + 1)

pcaResults <- prcomp(M)

pca <- autoplot(pcaResults, data = as.data.frame(annot), 
                colour = "dataset", shape = "phase",
                main = "Arbobios / CSBL datasets") +
  scale_color_manual(values = brewer.pal(3, "Dark2")[2:3]) +
  theme_bw()

ggsave(here("results","figures","PCA_Arbobios_CSBL_dataset.png"), pca, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2)
```

```{r Combat-seq}
df <- as.matrix(rawcount[,-1])
rownames(df) <- rawcount$Symbol

adj.count <- ComBat_seq(df, batch = pheno$dataset, group = pheno$condition)

out.adj.count <- adj.count %>% 
  as.data.frame() %>% 
  rownames_to_column('gene_id') %>% 
  full_join(hannot[,c(1,2,4)]) %>% 
  relocate('gene_id', GeneSymbol, Class)

#write_tsv(out.adj.count, here("data", "gene_expression", 
#                                "Arbobios_CSBL_rawCounts_CombatAdjusted.tsv"))

## x
expr.adj <- DGEList(counts = adj.count, 
                    genes = hannot[,c(1,2,4)],
                    group = paste(pheno$condition, pheno$phase, 
                                  pheno$disease.status, sep = "_"))

expr.adj.norm <- calcNormFactors(expr.adj, method = "TMM")

tmm.counts.adj <- cpm(expr.adj.norm, log = F)

out.tmm.counts.adj <- tmm.counts.adj %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  full_join(hannot %>% select('gene_id', "GeneSymbol", "Class")) %>% 
  relocate('gene_id', "GeneSymbol", "Class")

#write_tsv(out.tmm.counts.adj, 
#          here("data", "gene_expression",
#               "counts_Arbobios_CSBL_NOTmultimapping_TMM_allGenes.tsv"))

cpm.counts.adj <- cpm(expr.adj, log = F)

# Select most variant genes
genes.var.adj <- apply(tmm.counts.adj, 1, var)
selectedGenes.adj <- names(genes.var.adj[order(genes.var.adj, 
                                               decreasing = T)][1:500])

M.adj <- log2(t(tmm.counts.adj[selectedGenes.adj,]) + 1)

pcaResults.adj <- prcomp(M.adj)

pca.adj <- autoplot(pcaResults.adj, data = as.data.frame(annot), 
                colour = "dataset", shape = "phase",
                main = "Arbobios / CSBL datasets - Batch correction") +
  scale_color_manual(values = brewer.pal(3, "Dark2")[2:3]) +
  theme_bw()

ggsave(here("results", "figures",
            "PCA_Arbobios_CSBL_dataset_Combatseq_adjusted.png"), 
       pca.adj, device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2)

pca.adj.chron <- autoplot(pcaResults.adj, data = as.data.frame(annot), 
                          colour = "disease.status", shape = "phase",
                          main = "CHIKV Chronicity - Batch correction") +
  scale_colour_viridis(discrete = T) +
  theme_bw()

ggsave(here("results", "figures",
            "PCA_Arbobios_CSBL_dataset_Combatseq_adjusted_chronicity.png"), 
       pca.adj.chron, device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2)
```

