---
title: "Data integration"
editor_options: 
  chunk_output_type: console
---

```{r Load libraries}
library(here)
library(tidyverse)
library(edgeR)
library(mixOmics)
```

```{r Load files}
# CSBL CHIKV phenodata
sra.csbl <- read_csv(here("data", "chikv_csbl_star", "SraRunTable.txt"))

sup.csbl <- read_tsv(here("data", "chikv_csbl_star", "ppat.1007880.s001.csv"))

# Arbobios phenodata
pheno.arbo <- read_tsv(here("data", "phenodata.tsv"))

# Human genome annotation
hannot <- read_tsv(here("data", 
                        "Homo_sapiens.GRCh38.100_gene_annotation_table.txt"))

# CSBL read count
counts_csbl <- read_tsv(here("data", "chikv_csbl_star", "CountTable.tsv")) %>% 
  rename(Symbol = genes)

# Arbobobios read count
counts_arbo <- read_tsv(here("data", "geneCounts",
                          "rawCounts_featureCounts_NOTmultimapping.tsv")) %>%
  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
              ends_with("bam")) %>% 
  rename_with(~ sub("19017", "", .x), everything()) %>% 
  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>%
  rename_with(~ paste0("ARB", .x), -Symbol) %>%
  dplyr::select("Symbol", sort(colnames(.)))

total <- read_tsv(here("data", "gene_expression",
                       "counts_NOTmultimapping_TPM.tsv"))

micro <- read_tsv(here("data", "microRNA_sequencing", 
                       "mature_sense_minExpr1_RCadj_edger_TMMnormTable.csv"))
```

```{r Join CSBL and Arbobios phenodata + counts}
pheno.csbl <- sra.csbl %>% 
  full_join(sup.csbl, by = c("Sample Name" = "Subject_ID")) %>%
  mutate(condition = if_else(Sample_type == "Control", "healthy", 
                             "infected")) %>% 
  dplyr::select(Run, Label, Days_of_symptoms_Cat, Arthralgia_Chronicity,
                condition) %>% 
  mutate(day = if_else(Days_of_symptoms_Cat == "Control", NA_character_, 
                        Days_of_symptoms_Cat)) %>% 
  mutate(disease.status = case_when(Arthralgia_Chronicity == "ND" ~ NA_character_,
                             Arthralgia_Chronicity == "Y" ~ "chronic",
                             Arthralgia_Chronicity == "N" ~ "nonChronic"))

pheno <- pheno.arbo %>%
  rename(Run = Sample, day = Day) %>% 
  mutate(disease.status = ifelse(Class == "control", "nonChronic", Class)) %>% 
  mutate(condition = "infected") %>% 
  dplyr::select(-Class, - Batch) %>%
  mutate(Run = paste0("ARB", Run)) %>% 
  bind_rows(pheno.csbl %>% dplyr::select(Run, condition, day, 
                                         disease.status)) %>%
  mutate(phase = case_when(day == "D21" ~ "late",
                           day == "D20" ~ "late",
                           day == "D0" ~ "early",
                           day == "D0or1" ~ "early",
                           day == "D2" ~ "early",
                           day == "D3or4" ~ "early",
                           TRUE ~ NA_character_)) %>% 
  relocate(Run, condition, day, phase, disease.status) %>% 
  mutate(dataset = case_when(day == "D0" ~ "arbobios",
                           day == "D21" ~ "arbobios",
                           TRUE ~ "csbl")) %>% 
  mutate(condition = as.factor(condition)) %>% 
  mutate(dataset = as.factor(dataset))

rawcount <- counts_arbo %>% 
  full_join(counts_csbl, by = "Symbol") %>% 
  relocate(Symbol, pheno$Run)
```

```{r Normalization + PCA}
expr <- DGEList(counts = rawcount[,-1], 
                genes = hannot[,c(1,2,4)],
                group = paste(pheno$condition, pheno$phase, 
                              pheno$disease.status, sep = "_"))

expr <- calcNormFactors(expr, method = "TMM")

tmm.counts <- cpm(expr, log = F)
rownames(tmm.counts) <- rawcount$Symbol


annot <- pheno %>% 
  dplyr::select(Run, condition, phase, dataset) %>% 
  mutate(phase = ifelse(is.na(phase) == T, "healthy", phase)) %>% 
  column_to_rownames('Run')

# Select most variant genes
genes.var <- apply(tmm.counts, 1, var)
selectedGenes <- names(genes.var[order(genes.var, decreasing = T)][1:500])

M <- log2(t(tmm.counts[selectedGenes,]) + 1)

pcaResults <- prcomp(M)

pca <- autoplot(pcaResults, data = as.data.frame(annot), 
                colour = "dataset", shape = "condition") +
  scale_color_manual(values = brewer.pal(3, "Dark2")[2:3]) +
  theme_bw()

ggsave(here("results","figures","PCA_Arbobios_CSBL_dataset.png"), pca, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2)
```


```{r Combat-seq}
library(sva)

df <- as.matrix(rawcount[,-1])
rownames(df) <- rawcount$Symbol

phase <- ifelse(is.na(pheno$phase) == T, 'healthy', pheno$phase)

x <- ComBat_seq(df, pheno$dataset, pheno$condition)
y <- ComBat_seq(df, batch = pheno$dataset, group = NULL, 
                covar_mod = cbind(pheno$condition, as.factor(phase)))

## x
expr.adj <- DGEList(counts = x, 
                    genes = hannot[,c(1,2,4)],
                    group = paste(pheno$condition, pheno$phase, 
                                  pheno$disease.status, sep = "_"))

expr.adj <- calcNormFactors(expr.adj, method = "TMM")

tmm.counts.adj <- cpm(expr.adj, log = F)

# Select most variant genes
genes.var <- apply(tmm.counts.adj, 1, var)
selectedGenes <- names(genes.var[order(genes.var, decreasing = T)][1:500])

M <- log2(t(tmm.counts.adj[selectedGenes,]) + 1)

pcaResults <- prcomp(M)

autoplot(pcaResults, data = as.data.frame(annot), 
                colour = "phase", shape = "dataset") +
  scale_color_manual(values = brewer.pal(3, "Dark2")) +
  theme_bw()
```


```{r Pre-process}
# Rename colnames miRNA dataset
micro.rn <- micro %>% 
  rename_with(~ sub("X", "", .)) %>% 
  rename_with(~ sub("_trim", "", .)) %>% 
  column_to_rownames(var = "name")

# Select top variant genes from total RNA-seq dataset 
total.var <- apply(total[,-c(1:2)], 1, var)
names(total.var) <- total$Symbol

slctd.genes <- names(total.var[order(total.var, decreasing = T)])[1:5000]

# Filter top variant genes and select samples included in miRNA study
total.flt <- total %>% 
  filter(Symbol %in% slctd.genes) %>% 
  column_to_rownames(var = "GeneSymbol") %>% 
  dplyr::select(-Symbol) %>% 
  dplyr::select(colnames(micro.rn))

# List with transposed tibbles
data <- list(totalRNA = t(total.flt), miRNA = t(micro.rn))
```

