---
title: "RNA-seq analysis - CHIKV infection"
editor_options: 
  chunk_output_type: console
---

```{r Load libraries}
library(here)
library(tidyverse)
library(edgeR)
library(fgsea)

options(scipen=999)
```

```{r Load files}
# Human genome annotation
hannot <- read_tsv(here("data", 
                        "Homo_sapiens.GRCh38.100_gene_annotation_table.txt"))

# Phenodata
pheno <- read_tsv(here("data", "Arbobios_CSBL_CHIKV_phenodata.tsv"))

# CSBL read count Raw
raw_csbl <- read_tsv(here("data", "chikv_csbl_star", "CountTable.tsv")) %>% 
  dplyr::rename(Symbol = genes)

# Arbobios read count Raw
raw_arbobios <- read_tsv(here("data", "geneCounts",
                          "rawCounts_featureCounts_NOTmultimapping.tsv")) %>%
  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
              ends_with("bam")) %>% 
  rename_with(~ sub("19017", "", .x), everything()) %>% 
  rename_with(~ sub("^", "ARB", .x), -Symbol) %>%
  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>% 
  dplyr::select("Symbol", sort(colnames(.)))

# Read count adjusted by Combat-seq
adj.counts <- read_tsv(here("data", "gene_expression",
                            "Arbobios_CSBL_rawCounts_CombatAdjusted.tsv"))

# Reactome pathways
gene_set <- gmtPathways(here("data", "reactome_data", 
                             "ReactomePathwaysLevel3_2021-03.gmt"))
# Reactome root node
root_node <- read_tsv(here("data", "reactome_data", 
                           "ReactomePathways_leave_root_third_level.tsv"))
```

```{r DEGs Arbobios original data}
pheno_arb <- pheno %>% 
  filter(dataset == "arbobios")

group <- factor(paste(pheno_arb$phase, pheno_arb$disease.status, sep = "_"))

y <- DGEList(counts = raw_arbobios[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 1)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate Dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

my.contrast <- makeContrasts(
  chronVSnonChron.D0 = groupearly_chronic - groupearly_nonChronic,
  chronVSnonChron.D21 = grouplate_chronic - grouplate_nonChronic,
  chronD0VSchronD21 = groupearly_chronic - grouplate_chronic,
  nonChronD0VSnonChronD21 = groupearly_nonChronic - grouplate_nonChronic,
  levels = design)

chronVSnonChron.D0 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSnonChron.D0"])
chronVSnonChron.D21 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSnonChron.D21"])
chronD0VSchronD21 <- glmQLFTest(fit, contrast = my.contrast[, "chronD0VSchronD21"])
nonChronD0VSnonChronD21 <- glmQLFTest(fit, 
                                      contrast = my.contrast[, "nonChronD0VSnonChronD21"])

chronVSnonChron.D0_t <- topTags(chronVSnonChron.D0, n = NULL)$table
chronVSnonChron.D21_t <- topTags(chronVSnonChron.D21, n = NULL)$table
chronD0VSchronD21_t <- topTags(chronD0VSchronD21, n = NULL)$table
nonChronD0VSnonChronD21_t <- topTags(nonChronD0VSnonChronD21, n = NULL)$table

################## DEGs by Class
group <- factor(pheno_arb$disease.status)

y <- DGEList(counts = raw_arbobios[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 1)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

chronVSnonChron <- glmQLFTest(fit, contrast = c(1,-1))
chronVSnonChron_t <- topTags(chronVSnonChron, n = NULL)$table

################## DEGs by Day
group <- factor(pheno_arb$phase)

y <- DGEList(counts = raw_arbobios[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 1)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

D0VSD21 <- glmQLFTest(fit, contrast = c(1,-1))
D0VSD21_t <- topTags(D0VSD21, n = NULL)$table

edgeR.arb_original <- bind_rows("chronVSnonChron.D0" = chronVSnonChron.D0_t,
                       "chronVSnonChron.D21" = chronVSnonChron.D21_t,
                       "chronD0VSchronD21" = chronD0VSchronD21_t,
                       "nonChronD0VSnonChronD21" = nonChronD0VSnonChronD21_t,
                       "chronVSnonChron" = chronVSnonChron_t,
                       "D0VSD21" = D0VSD21_t,
                      .id = "compar")

write_tsv(edgeR.arb_original, 
          here("results", "DEG_edgeR", 
               "arbobios_phase_condition_original.tsv"))

fc.up = 1
fc.dw = -1
stat = 0.1

edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D0") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D0") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()  

edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow() 

edgeR.arb_original %>%
  filter(compar == "chronD0VSchronD21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "chronD0VSchronD21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()

edgeR.arb_original %>%
  filter(compar == "nonChronD0VSnonChronD21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "nonChronD0VSnonChronD21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()

edgeR.arb_original %>%
  filter(compar == "chronVSnonChron") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "chronVSnonChron") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()

edgeR.arb_original %>%
  filter(compar == "D0VSD21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "D0VSD21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()
```

```{r DEGs CSBL original data}
annot <- pheno %>% 
  dplyr::select(Run, condition, phase, dataset, disease.status) %>% 
  mutate(phase = ifelse(is.na(phase) == T, "healthy", phase)) %>% 
  mutate(disease.status = case_when(phase == "healthy" ~ "healthy", 
                                    phase != 'healthy' &
                                    is.na(disease.status) ~ 'unknown',
                                    TRUE ~ disease.status)) %>% 
  column_to_rownames('Run')

csbl.annot <- annot %>% 
  filter(dataset == "csbl") %>% 
  mutate(group = ifelse(condition == "healthy", "healthy", 
                        paste(condition, phase, disease.status, sep = '_'))) %>% 
  rownames_to_column("Run")

####### Phase comparison - Early VS Healthy
groups <- as.factor(csbl.annot$phase)

counts_csbl_ord <- raw_csbl %>% 
  dplyr::select(Symbol, csbl.annot$Run)

csbl.expr <- DGEList(counts = counts_csbl_ord[,-1], 
                     genes = hannot[,c(1,2,4)],
                     group = groups)

design <- model.matrix(~groups+0)

# Filtering low expression genes
keep <- filterByExpr(csbl.expr, min.count = 1)
csbl.expr <- csbl.expr[keep, , keep.lib.sizes=FALSE]

# Norm factors
csbl.expr <- calcNormFactors(csbl.expr, method = "TMM")

# Estimate Dispersion
csbl.expr <- estimateDisp(csbl.expr, design)

#DE
fit <- glmQLFit(csbl.expr, design)

my.contrast <- makeContrasts(earlyVShealthy = groupsearly - groupshealthy,
                            levels = design)

earlyVShealthy <- glmQLFTest(fit, contrast = my.contrast[, "earlyVShealthy"])

earlyVShealthy_t <- topTags(earlyVShealthy, n = NULL)$table

####### Chronicity comparison - Early Chronic VS Early Non-chronic
groups <- as.factor(csbl.annot$group)

csbl.expr <- DGEList(counts = counts_csbl_ord[,-1], 
                     genes = hannot[,c(1,2,4)],
                     group = groups)

design <- model.matrix(~groups+0)

# Filtering low expression genes
keep <- filterByExpr(csbl.expr, min.count = 1)
csbl.expr <- csbl.expr[keep, , keep.lib.sizes=FALSE]

# Norm factors
csbl.expr <- calcNormFactors(csbl.expr, method = "TMM")

# Estimate Dispersion
csbl.expr <- estimateDisp(csbl.expr, design)

#DE
fit <- glmQLFit(csbl.expr, design)

my.contrast <- makeContrasts(
  chronicVSnonChronic.early = groupsinfected_early_chronic - 
                              groupsinfected_early_nonChronic,
  levels = design)

chronVSnonChron.early <- glmQLFTest(fit, contrast = my.contrast[, "chronicVSnonChronic.early"])

chronVSnonChron.early_t <- topTags(chronVSnonChron.early, n = NULL)$table

### DEGs number
# Chron vs Non-Chron
chronVSnonChron.early_t %>% 
  filter(logFC >= 1, FDR < 0.1) %>% nrow()
chronVSnonChron.early_t %>% 
  filter(logFC <= -1, FDR < 0.1) %>% nrow()

# Early vs Late
earlyVShealthy_t %>% 
   filter(logFC >= 1, FDR < 0.1) %>% nrow()
earlyVShealthy_t %>% 
  filter(logFC <= -1, FDR < 0.1) %>% nrow()

edgeR.out <- bind_rows("earlyVShealthy" = earlyVShealthy_t, 
                      "chronVSnonChron.early" = chronVSnonChron.early_t,
                      .id = "compar")

#write_tsv(edgeR.out, here("results", "DEG_edgeR", "csbl_phase_condition.tsv"))
```

```{r DEGs Arbobios adjusted data}
pheno_arb <- pheno %>% 
  filter(dataset == "arbobios")

group <- factor(paste(pheno_arb$phase, pheno_arb$disease.status, sep = "_"))

raw_arbobios <- adj.counts %>% 
  dplyr::select(Symbol = gene_id, pheno_arb$Run)

y <- DGEList(counts = raw_arbobios[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 5)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate Dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

my.contrast <- makeContrasts(
  chronVSnonChron.D0 = groupearly_chronic - groupearly_nonChronic,
  chronVSnonChron.D21 = grouplate_chronic - grouplate_nonChronic,
  chronD0VSchronD21 = groupearly_chronic - grouplate_chronic,
  nonChronD0VSnonChronD21 = groupearly_nonChronic - grouplate_nonChronic,
  levels = design)

chronVSnonChron.D0 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSnonChron.D0"])
chronVSnonChron.D21 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSnonChron.D21"])
chronD0VSchronD21 <- glmQLFTest(fit, contrast = my.contrast[, "chronD0VSchronD21"])
nonChronD0VSnonChronD21 <- glmQLFTest(fit, 
                                      contrast = my.contrast[, "nonChronD0VSnonChronD21"])

chronVSnonChron.D0_t <- topTags(chronVSnonChron.D0, n = NULL)$table
chronVSnonChron.D21_t <- topTags(chronVSnonChron.D21, n = NULL)$table
chronD0VSchronD21_t <- topTags(chronD0VSchronD21, n = NULL)$table
nonChronD0VSnonChronD21_t <- topTags(nonChronD0VSnonChronD21, n = NULL)$table

################## DEGs by Class
group <- factor(pheno_arb$disease.status)

y <- DGEList(counts = raw_arbobios[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 5)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

chronVSnonChron <- glmQLFTest(fit, contrast = c(1,-1))
chronVSnonChron_t <- topTags(chronVSnonChron, n = NULL)$table

################## DEGs by Day
group <- factor(pheno_arb$phase)

y <- DGEList(counts = raw_arbobios[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 5)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

D0VSD21 <- glmQLFTest(fit, contrast = c(1,-1))
D0VSD21_t <- topTags(D0VSD21, n = NULL)$table

edgeR.arb_original <- bind_rows("chronVSnonChron.D0" = chronVSnonChron.D0_t,
                       "chronVSnonChron.D21" = chronVSnonChron.D21_t,
                       "chronD0VSchronD21" = chronD0VSchronD21_t,
                       "nonChronD0VSnonChronD21" = nonChronD0VSnonChronD21_t,
                       "chronVSnonChron" = chronVSnonChron_t,
                       "D0VSD21" = D0VSD21_t,
                      .id = "compar")

#write_tsv(edgeR.arb_original, 
#          here("results", "DEG_edgeR", 
#               "arbobios_phase_condition_original.tsv"))

fc.up = 1
fc.dw = -1
stat = 0.05

edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D0") %>% 
  filter(logFC > fc.up & FDR < stat) 
edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D0") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()  

edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "chronVSnonChron.D21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow() 

edgeR.arb_original %>%
  filter(compar == "chronD0VSchronD21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "chronD0VSchronD21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()

edgeR.arb_original %>%
  filter(compar == "nonChronD0VSnonChronD21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "nonChronD0VSnonChronD21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()

edgeR.arb_original %>%
  filter(compar == "chronVSnonChron") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "chronVSnonChron") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()

edgeR.arb_original %>%
  filter(compar == "D0VSD21") %>% 
  filter(logFC > fc.up & FDR < stat) %>% nrow()
edgeR.arb_original %>%
  filter(compar == "D0VSD21") %>% 
  filter(logFC < fc.dw & FDR < stat) %>% nrow()
```

```{r DEGs CSBL adjusted data}
annot <- pheno %>% 
  dplyr::select(Run, condition, phase, dataset, disease.status) %>% 
  mutate(phase = ifelse(is.na(phase) == T, "healthy", phase)) %>% 
  mutate(disease.status = case_when(phase == "healthy" ~ "healthy", 
                                    phase != 'healthy' &
                                    is.na(disease.status) ~ 'unknown',
                                    TRUE ~ disease.status)) %>% 
  column_to_rownames('Run')

csbl.annot <- annot %>% 
  filter(dataset == "csbl") %>% 
  mutate(group = ifelse(condition == "healthy", "healthy", 
                        paste(condition, phase, disease.status, sep = '_'))) %>% 
  rownames_to_column("Run")

####### Phase comparison - Early VS Healthy
groups <- as.factor(csbl.annot$phase)

counts_csbl_ord <- adj.counts %>% 
  dplyr::select(Symbol = gene_id, csbl.annot$Run)

csbl.expr <- DGEList(counts = counts_csbl_ord[,-1], 
                     genes = hannot[,c(1,2,4)],
                     group = groups)

design <- model.matrix(~groups+0)

# Filtering low expression genes
keep <- filterByExpr(csbl.expr, min.count = 5)
csbl.expr <- csbl.expr[keep, , keep.lib.sizes=FALSE]

# Norm factors
csbl.expr <- calcNormFactors(csbl.expr, method = "TMM")

# Estimate Dispersion
csbl.expr <- estimateDisp(csbl.expr, design)

#DE
fit <- glmQLFit(csbl.expr, design)

my.contrast <- makeContrasts(earlyVShealthy = groupsearly - groupshealthy,
                            levels = design)

earlyVShealthy <- glmQLFTest(fit, contrast = my.contrast[, "earlyVShealthy"])

earlyVShealthy_t <- topTags(earlyVShealthy, n = NULL)$table

####### Chronicity comparison - Early Chronic VS Early Non-chronic
groups <- as.factor(csbl.annot$group)

csbl.expr <- DGEList(counts = counts_csbl_ord[,-1], 
                     genes = hannot[,c(1,2,4)],
                     group = groups)

design <- model.matrix(~groups+0)

# Filtering low expression genes
keep <- filterByExpr(csbl.expr, min.count = 5)
csbl.expr <- csbl.expr[keep, , keep.lib.sizes=FALSE]

# Norm factors
csbl.expr <- calcNormFactors(csbl.expr, method = "TMM")

# Estimate Dispersion
csbl.expr <- estimateDisp(csbl.expr, design)

#DE
fit <- glmQLFit(csbl.expr, design)

my.contrast <- makeContrasts(
  chronicVSnonChronic.early = groupsinfected_early_chronic - 
                              groupsinfected_early_nonChronic,
  levels = design)

chronVSnonChron.early <- glmQLFTest(fit, contrast = my.contrast[, "chronicVSnonChronic.early"])

chronVSnonChron.early_t <- topTags(chronVSnonChron.early, n = NULL)$table

### DEGs number
# Chron vs Non-Chron
chronVSnonChron.early_t %>% 
  filter(logFC >= 1, FDR < 0.05) %>% nrow()
chronVSnonChron.early_t %>% 
  filter(logFC <= -1, FDR < 0.05) %>% nrow()

# Early vs Late
earlyVShealthy_t %>% 
   filter(logFC >= 1, FDR < 0.05) %>% nrow()
earlyVShealthy_t %>% 
  filter(logFC <= -1, FDR < 0.05) %>% nrow()

edgeR.out <- bind_rows("earlyVShealthy" = earlyVShealthy_t, 
                      "chronVSnonChron.early" = chronVSnonChron.early_t,
                    .id = "compar")
```

```{r DEGs Arbobios + CSBL adjusted data}
sub.pheno <- pheno %>% 
  mutate(group = paste(phase, disease.status, sep = "_")) %>% 
  mutate(group = ifelse(group == "NA_NA", "healthy", group)) %>% 
  filter(group != "early_NA") %>% 
  filter(!(dataset == "csbl" & group != "healthy")) %>% 
  mutate(group = as.factor(group)) %>% 
  mutate(phase = ifelse(is.na(phase), "healthy", phase))

group <- sub.pheno$group

adj_count <- adj.counts %>% 
  dplyr::select(Symbol = gene_id, sub.pheno$Run)

y <- DGEList(counts = adj_count[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 5)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate Dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

my.contrast <- makeContrasts(
  nonChron.D0VShealthy = groupearly_nonChronic - grouphealthy,
  nonChron.D21VShealthy = grouplate_nonChronic - grouphealthy,
  chron.D0VShealthy =  groupearly_chronic - grouphealthy,
  chron.D21VShealthy = grouplate_chronic - grouphealthy,
  levels = design)

nonChron.D0VShealthy <- glmQLFTest(fit, contrast = my.contrast[, "nonChron.D0VShealthy"])
nonChron.D21VShealthy <- glmQLFTest(fit, contrast = my.contrast[, "nonChron.D21VShealthy"])
chron.D0VShealthy <- glmQLFTest(fit, contrast = my.contrast[, "chron.D0VShealthy"])
chron.D21VShealthy <- glmQLFTest(fit, contrast = my.contrast[, "chron.D21VShealthy"])

nonChron.D0VShealthy_t <- topTags(nonChron.D0VShealthy, n = NULL)$table
nonChron.D21VShealthy_t <- topTags(nonChron.D21VShealthy, n = NULL)$table
chron.D0VShealthy_t <- topTags(chron.D0VShealthy, n = NULL)$table
chron.D21VShealthy_t <- topTags(chron.D21VShealthy, n = NULL)$table

################## DEGs by Day
sub.pheno <- pheno %>% 
  mutate(phase = ifelse(is.na(phase), "healthy", phase)) 

group <- factor(sub.pheno$phase)

adj_count <- adj.counts %>% 
  dplyr::select(Symbol = gene_id, sub.pheno$Run)

y <- DGEList(counts = adj_count[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 5)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)

# Estimate Dispersion
y <- estimateDisp(y, design)

#DE
fit <- glmQLFit(y, design)

my.contrast <- makeContrasts(
  earlyVShealthy = groupearly - grouphealthy,
  lateVShealthy = grouplate - grouphealthy,
  levels = design)

earlyVShealthy <- glmQLFTest(fit, contrast = my.contrast[, "earlyVShealthy"])
lateVShealthy <- glmQLFTest(fit, contrast = my.contrast[, "lateVShealthy"])

earlyVShealthy_t <- topTags(earlyVShealthy, n = NULL)$table
lateVShealthy_t <- topTags(lateVShealthy, n = NULL)$table

edgeR.arb.csbl <- bind_rows("nonChron.D0VShealthy" = nonChron.D0VShealthy_t,
                            "nonChron.D21VShealthy" = nonChron.D21VShealthy_t,
                            "chron.D0VShealthy" = chron.D0VShealthy_t,
                            "chron.D21VShealthy" = chron.D21VShealthy_t,
                            "earlyVShealthy" = earlyVShealthy_t,
                            "lateVShealthy" = lateVShealthy_t,
                            .id = "compar")

## Number DEGs
nonChron.D0VShealthy_t %>% filter(logFC >= 1, FDR < 0.01) %>% nrow()
nonChron.D0VShealthy_t %>% filter(logFC <= -1, FDR < 0.01) %>% nrow()
nonChron.D21VShealthy_t %>% filter(logFC >= 1, FDR < 0.01) %>% nrow()
nonChron.D21VShealthy_t %>% filter(logFC <= -1, FDR < 0.01) %>% nrow()
chron.D0VShealthy_t %>% filter(logFC >= 1, FDR < 0.01) %>% nrow()
chron.D0VShealthy_t %>% filter(logFC <= -1, FDR < 0.01) %>% nrow()
chron.D21VShealthy_t %>% filter(logFC >= 1, FDR < 0.01) %>% nrow()
chron.D21VShealthy_t %>% filter(logFC <= -1, FDR < 0.01) %>% nrow()

earlyVShealthy_t %>% filter(logFC >= 1, FDR < 0.01) %>% nrow()
earlyVShealthy_t %>% filter(logFC <= -1, FDR < 0.01) %>% nrow()
lateVShealthy_t %>% filter(logFC >= 1, FDR < 0.01) %>% nrow()
lateVShealthy_t %>% filter(logFC <= -1, FDR < 0.01) %>% nrow()

## Union

nonChron.D0.up <- nonChron.D0VShealthy_t %>% filter(logFC >= 1, FDR < 0.01)
nonChron.D0.down <- nonChron.D0VShealthy_t %>% filter(logFC <= -1, FDR < 0.01)
nonChron.D21.up <- nonChron.D21VShealthy_t %>% filter(logFC >= 1, FDR < 0.01)
nonChron.D21.down <- nonChron.D21VShealthy_t %>% filter(logFC <= -1, FDR < 0.01)

chron.D0.up <- chron.D0VShealthy_t %>% filter(logFC >= 1, FDR < 0.01)
chron.D0.down <-chron.D0VShealthy_t %>% filter(logFC <= -1, FDR < 0.01)
chron.D21.up <-chron.D21VShealthy_t %>% filter(logFC >= 1, FDR < 0.01)
chron.D21.down <-chron.D21VShealthy_t %>% filter(logFC <= -1, FDR < 0.01)


sum(chron.D0.up$GeneSymbol %in% nonChron.D0.up$GeneSymbol)
sum(chron.D0.down$GeneSymbol %in% nonChron.D0.down$GeneSymbol)

```

```{r fGSEA}
#nonChron D0
ranks_nonChron.D0VShealthy <-  nonChron.D0VShealthy_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_nonChron.D0VShealthy <- fgsea(pathways=gene_set, 
                                       stats=ranks_nonChron.D0VShealthy, eps=0)

#nonChron D21
ranks_nonChron.D21VShealthy <-  nonChron.D21VShealthy_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_nonChron.D21VShealthy <- fgsea(pathways=gene_set, 
                                       stats=ranks_nonChron.D21VShealthy, eps=0)

#chron D0
ranks_chron.D0VShealthy <- chron.D0VShealthy_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chron.D0VShealthy <- fgsea(pathways=gene_set, 
                                       stats=ranks_chron.D0VShealthy, eps=0)
#chron D21
ranks_chron.D21VShealthy <- chron.D21VShealthy_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chron.D21VShealthy <- fgsea(pathways=gene_set, 
                                       stats=ranks_chron.D21VShealthy, eps=0)
# D0
ranks_earlyVShealthy <- earlyVShealthy_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_earlyVShealthy <- fgsea(pathways=gene_set, 
                                       stats=ranks_earlyVShealthy, eps=0)
# D21
ranks_lateVShealthy <- lateVShealthy_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_lateVShealthy <- fgsea(pathways=gene_set, 
                                       stats=ranks_lateVShealthy, eps=0)

## Join results
fgseaRes <- bind_rows("nonChron.D0VShealthy" = fgseaRes_nonChron.D0VShealthy, 
                      "nonChron.D21VShealthy" = fgseaRes_nonChron.D21VShealthy,
                      "chron.D0VShealthy" = fgseaRes_chron.D0VShealthy,
                      "chron.D21VShealthy" = fgseaRes_chron.D21VShealthy,
                      "earlyVShealthy" = fgseaRes_earlyVShealthy,
                      "lateVShealthy" = fgseaRes_lateVShealthy, 
                      .id = "compar") %>% 
  left_join(root_node %>% 
              dplyr::select(root_name, node_name), 
            by = c("pathway" = "node_name")) %>% 
  relocate(compar, pathway, root_name)

## Plot
custom_pal <- c("#053061", "#2166AC", "#4393C3", "#92C5DE",
                        "#D1E5F0", "#FFFFFF", "#FDDBC7", "#F4A582",
                        "#D6604D", "#B2182B", "#67001F")
custom_pal <- colorRampPalette(custom_pal)(200)

fgseaRes_plot <- fgseaRes %>% 
  group_by(pathway) %>% 
  filter(any(padj < 1e-3)) %>% 
  mutate(NES = ifelse(padj > 1e-3, 0, NES)) %>% 
  na.omit() %>% 
  mutate(compar = factor(compar, 
                         levels = c("nonChron.D0VShealthy",
                                    "chron.D0VShealthy", 
                                    "nonChron.D21VShealthy",
                                    "chron.D21VShealthy",
                                    "earlyVShealthy", "lateVShealthy")))

max_abs_nes <- max(abs(fgseaRes_plot$NES))

fgseaRes_plot %>% 
  ggplot(aes(compar, y=pathway, size=-log10(padj), fill=NES)) +
  geom_point(color = 'white', shape=21) +
  scale_fill_gradientn(colours=custom_pal, space = "Lab",
                       limits=c(-max_abs_nes, max_abs_nes),
                       ) +
  facet_grid(rows = vars(root_name), scales = "free_y", space = "free_y",
             switch = "y") +
  guides(size = guide_legend(override.aes = list(fill = "black")))+
  theme_bw() +
  theme(panel.grid.major = element_line(linetype = "dashed"),
        panel.spacing = unit(0, "lines"),
        axis.text.x = element_text(angle = 30, vjust = 1, hjust=1),
        strip.text.y.left = element_text(angle = 0),
        strip.background =element_rect(fill=NA))
```

