---
title: "RNA-seq analysis - CHIKV infection"
editor_options: 
  chunk_output_type: console
---

```{r Load libraries}
library(here)
library(tidyverse)
library(edgeR)
library(EDASeq)
library(pheatmap)
library(ggfortify)
library(RColorBrewer)
library(fgsea)

options(scipen=999)
```

```{r Load files}
# Immune response genes
immgene <- read_tsv(here("data", "immune_response_GO0006955_genes.txt"),
                    col_names = F)

# Human genome annotation
hannot <- read_tsv(here("data", 
                        "Homo_sapiens.GRCh38.100_gene_annotation_table.txt"))

# Gene length
gl <- read_tsv(here("data", "GC_lengths.tsv"), col_names = F, skip = 1) %>% 
  dplyr::rename(gene_id = X1, gene_length = X2) %>% 
  dplyr::select(-X3)
  
# Phenodata
pheno <- read_tsv(here("data", "phenodata.tsv")) %>% 
  mutate(Class = factor(Class), Day  = factor(Day)) 
  
# Read not multimapping file
rawcount_notmulti <- read_tsv(here("data", "geneCounts",
                          "rawCounts_featureCounts_NOTmultimapping.tsv")) %>%
  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
              ends_with("bam")) %>% 
  rename_with(~ sub("19017", "", .x), everything()) %>% 
  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>% 
  dplyr::select("Symbol", sort(colnames(.)))

# Read multimapping file
#rawcount_multi <- read_tsv(here("data", "geneCounts",
#                          "rawCounts_featureCounts_multimapping.tsv")) %>%
#  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
#              ends_with("bam")) %>% 
#  rename_with(~ sub("19017", "", .x), everything()) %>% 
#  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>% 
#  dplyr::select("Symbol", sort(colnames(.)))
  
# Reactome gene set
#gene_set <- read_tsv(here("data", "reactome_data", 
#                          "ReactomePathwaysLevel3_ensembl_2021-02.tsv"))

gene_set <- gmtPathways(here("data", "reactome_data", 
                             "ReactomePathwaysLevel3_2021-03.gmt"))

# Read file with 75 interferon and interleukine genes
ilifn <- read_tsv(here("data", "IFN_IL_list.txt"))
```

```{r Alignment Stat}
# feature count stat not multimapping
stat_notmulti <- read_tsv(here("data", "geneCounts", 
                               "featureCounts_stat_NOTmultimapping.tsv")) %>%
  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
              ends_with("bam")) %>% 
  rename_with(~ sub("19017", "", .x), everything()) %>% 
  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>% 
  pivot_longer(-Status, names_to = "lib", values_to = "reads") %>% 
  mutate(protocol = "NotMulti")

stat_multi <- read_tsv(here("data", "geneCounts", 
                               "featureCounts_stat_multimapping.tsv")) %>%
  rename_with(~ sub(".Aligned.sortedByCoord.out.bam", "", .x), 
              ends_with("bam")) %>% 
  rename_with(~ sub("19017", "", .x), everything()) %>% 
  rename_with(~ sub("Aligned.out.bam", "", .x), everything()) %>% 
  pivot_longer(-Status, names_to = "lib", values_to = "reads") %>% 
  mutate(protocol = "Multi")

x <- stat_notmulti %>% 
  bind_rows(stat_multi) %>% 
  filter(Status == "Assigned") %>% 
  ggplot(aes(x = lib, y = reads, fill = protocol)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  scale_fill_manual(values = brewer.pal(3,"RdYlBu")[c(1,3)]) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=4),
        panel.grid.major.x = element_blank())+
  scale_y_continuous(labels = scales::number)

ggsave(here("results","figures", "featureCount_stats.png"), plotstat, 
       device = "png", width = 12, height = 7, units = "cm", 
       dpi = "print", scale = 2)

```

```{r Normalization}
# Not multimapping TPM
countnotmulti <- rawcount_notmulti %>%
  full_join(hannot[,1:2], by = c('Symbol' = 'gene_id')) %>% 
  inner_join(gl, by = c('Symbol' = 'gene_id')) %>% 
  relocate(Symbol, GeneSymbol, gene_length)

TPMnotmulti <- scater::calculateTPM(countnotmulti[,-c(1:3)], 
                                    countnotmulti$gene_length)

outTPMnotmulti <- TPMnotmulti %>% 
  as_tibble() %>% 
  mutate(Symbol = countnotmulti$Symbol, 
         GeneSymbol = countnotmulti$GeneSymbol) %>% 
  relocate(Symbol, GeneSymbol)

#write_tsv(outTPMnotmulti,
#          here("data", "gene_expression", "counts_NOTmultimapping_TPM.tsv"))

# Not multimapping CPM + TMM
group <- factor(paste(pheno$Class, pheno$Day, sep = "_"))

expr <- DGEList(counts = rawcount_notmulti[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

expr.norm.fact <- calcNormFactors(expr, method = 'TMM')

tmm.counts <- cpm(expr.norm.fact, log = F)
out.tmm.counts <- cbind(hannot[,c(1,2,4)], tmm.counts)

#write_tsv(out.tmm.counts,
#          here("data", "gene_expression", 
#               "counts_NOTmultimapping_TMM_allGenes.tsv"))

cpm.counts <- cpm(expr, log = F)
out.cpm.counts <- cbind(hannot[,1:2], cpm.counts)

#write_tsv(out.cpm.counts,
#          here("data", "gene_expression", "counts_NOTmultimapping_CPM.tsv"))

# Multimapping
#countmulti <- rawcount_multi %>%
#  full_join(hannot[,1:2], by = c('Symbol' = 'gene_id')) %>%
#  inner_join(gl, by = c('Symbol' = 'gene_id')) %>% 
#  relocate(Symbol, GeneSymbol, gene_length)

#TPMmulti <- scater::calculateTPM(countmulti[,-c(1:3)], countmulti$gene_length)

#outTPMmulti <- TPMmulti %>% 
#  as_tibble() %>% 
#  mutate(Symbol = countmulti$Symbol, 
#         GeneSymbol = countmulti$GeneSymbol) %>% 
#  relocate(Symbol, GeneSymbol)

#write_tsv(outTPMmulti,
#          here("data", "gene_expression", "counts_multimapping_TPM.tsv"))

```

```{r Heatmap}
# Not multimapping
normcount_notmulti <- edgeR::cpm(rawcount_notmulti[,-1])
rownames(normcount_notmulti) <- rawcount_notmulti$Symbol

V_notmulti <- apply(normcount_notmulti, 1, var)

selectedGenes_notmulti <- names(V_notmulti[order(V_notmulti, 
                                                 decreasing = T)][1:500])

annot <- pheno[,2:4]
rownames(annot) <- pheno$Sample

heatmap_notmulti <- pheatmap(normcount_notmulti[selectedGenes_notmulti,], 
                             scale = 'row', show_rownames = F, 
                             annotation_col = as.data.frame(annot),
                             show_colnames = F)

ggsave(here("results","figures","Heatmap_notmulti.png"), heatmap_notmulti, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2)

# Multimapping
#normcount_multi <- edgeR::cpm(rawcount_multi[,-1])
#rownames(normcount_multi) <- rawcount_multi$Symbol

#V_multi <- apply(normcount_multi, 1, var)

#selectedGenes_multi <- names(V_multi[order(V_multi, decreasing = T)][1:500])

#heatmap_multi <- pheatmap(normcount_multi[selectedGenes_multi,], 
#                          scale = 'row', show_rownames = F, 
#                          annotation_col = as.data.frame(annot),
#                          show_colnames = F)

#ggsave(here("results","figures","Heatmap_multi.png"), heatmap_multi, 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 2)
```

```{r RLE}
colors <- brewer.pal(3, "Set2")

# Not multimapping
plotRLE(as.matrix(rawcount_notmulti[,-1]), outline = FALSE, 
        col = colors[pheno$Class], ylim = c(-4, 4), 
        main = "Raw counts - Not multimapping", 
        ylab= "Relative Log Expression")

plotRLE(cpm(rawcount_notmulti[,-1]), outline=FALSE, col = colors[pheno$Class],
        ylim = c(-4, 4), main = "CPM - Not multimapping", 
        ylab= "Relative Log Expression")

# Multimapping
#plotRLE(as.matrix(rawcount_multi[,-1]), outline = FALSE, 
#        col = colors[pheno$Class], ylim = c(-4, 4), 
#        main = "Raw counts - Multimapping", ylab= "Relative Log Expression")

#plotRLE(cpm(rawcount_multi[,-1]), outline=FALSE, col = colors[pheno$Class],
#        ylim = c(-4, 4), main = "CPM - Multimapping", 
#        ylab= "Relative Log Expression")
```

```{r PCA}
# Not multimapping CPM
M_notmulti <- t(normcount_notmulti[selectedGenes_notmulti,])

M_notmulti <- log2(M_notmulti + 1)

pcaResults_notmulti <- prcomp(M_notmulti)

pca_notmulti <- autoplot(pcaResults_notmulti, data = as.data.frame(annot), 
         colour = 'Day', shape = "Class") +
  scale_color_manual(values = brewer.pal(3, "Dark2")[2:3]) +
  theme_bw()

ggsave(here("results","figures","PCA_notmulti_CPM.png"), pca_notmulti, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 1)

# Not multimapping TMM
rownames(out.tmm.counts) <- out.tmm.counts$gene_id
  
V_notmulti_tmm <- apply(out.tmm.counts[,-c(1:2)], 1, var)

selectedGenes_notmulti_tmm <- names(V_notmulti_tmm[order(V_notmulti_tmm, 
                                                       decreasing = T)][1:500])


M_notmulti_tmm <- t(out.tmm.counts[selectedGenes_notmulti_tmm, -c(1:2)])

M_notmulti_tmm <- log2(M_notmulti_tmm + 1)

pcaResults_notmulti_tmm <- prcomp(M_notmulti_tmm)

pca_notmulti_tmm <- autoplot(pcaResults_notmulti_tmm, data = as.data.frame(annot), 
         colour = 'Day', shape = "Class", main = "NotMulti TMM") +
  scale_color_manual(values = brewer.pal(3, "Dark2")[2:3]) +
  theme_bw()

ggsave(here("results","figures","PCA_notmulti_TMM.png"), pca_notmulti_tmm, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 1)

# Multimapping
#M_multi <- t(normcount_multi[selectedGenes_multi,])

#M_multi <- log2(M_multi + 1)

#pcaResults_multi <- prcomp(M_multi)

#pca_multi <- autoplot(pcaResults_multi, data = as.data.frame(annot), 
#         colour = 'Day', shape = "Class") +
#  scale_color_manual(values = brewer.pal(3, "Dark2")[2:3]) +
#  theme_bw()

#ggsave(here("results","figures","PCA_multi_CPM.png"), pca_multi, 
#       device = "png", width = 10, height = 6, units = "cm", 
#       dpi = "print", scale = 1)
```

```{r DE analysis Not multimapping - edgeR}
group <- factor(paste(pheno$Class, pheno$Day, sep = "_"))

#Not multimapping
y <- DGEList(counts = rawcount_notmulti[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 5)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)
#plotMDS(y, col = brewer.pal(4,"Set2")[group], labels = group)

tmm <- cpm(y)
rownames(tmm) <- y$genes$gene_id

out.tmm <- tmm %>% 
  as.data.frame() %>% 
  rownames_to_column('gene_id') %>% 
  left_join(hannot[,c(1,2,4)]) %>% 
  relocate("gene_id", "GeneSymbol", "Class")

#write_tsv(out.tmm,
#          here("data", "gene_expression", 
#               "counts_NOTmultimapping_TMM_filtered_20k.tsv"))

# Estimate Dispersion
y <- estimateDisp(y, design)
#plotBCV(y)

#DE
fit <- glmQLFit(y, design)

my.contrast <- makeContrasts(
  chronVSctrl.D0 = groupchronic_D0 - groupcontrol_D0,
  chronVSctrl.D21 = groupchronic_D21 - groupcontrol_D21,
  chronD21VSchronD0 = groupchronic_D21 - groupchronic_D0,
  ctrlD21VSctrlD0 = groupcontrol_D21 - groupcontrol_D0,
  chronVSctrl = (groupchronic_D21 + groupchronic_D0) - (groupcontrol_D21 + groupcontrol_D0),
  D21VSD0 = (groupchronic_D21 + groupcontrol_D21) - (groupchronic_D0 + groupcontrol_D0),
  levels = design)

chronVSctrl.D0 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSctrl.D0"])
chronVSctrl.D21 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSctrl.D21"])
chronD21VSchronD0 <- glmQLFTest(fit, contrast = my.contrast[, "chronD21VSchronD0"])
ctrlD21VSctrlD0 <- glmQLFTest(fit, contrast = my.contrast[, "ctrlD21VSctrlD0"])
chronVSctrl <- glmQLFTest(fit, contrast = my.contrast[, "chronVSctrl"])
D21VSD0 <- glmQLFTest(fit, contrast = my.contrast[, "D21VSD0"])

chronVSctrl.D0_t <- topTags(chronVSctrl.D0, n = NULL)$table
chronVSctrl.D21_t <- topTags(chronVSctrl.D21, n = NULL)$table
chronD21VSchronD0_t <- topTags(chronD21VSchronD0, n = NULL)$table
ctrlD21VSctrlD0_t <- topTags(ctrlD21VSctrlD0, n = NULL)$table
chronVSctrl_t <- topTags(chronVSctrl, n = NULL)$table
D21VSD0_t <- topTags(D21VSD0, n = NULL)$table

meanTMM <- cbind(y$genes[,1:3],
                 rowMeans(tmm[,group == 'control_D0']), 
                 rowMeans(tmm[,group == 'control_D21']),
                 rowMeans(tmm[,group == 'chronic_D0']),
                 rowMeans(tmm[,group == 'chronic_D21']))

colnames(meanTMM) <- c("gene_id", "GeneSymbol", "Class",
                       'control_D0_TMM', 'control_D21_TMM', 
                       'chronic_D0_TMM', 'chronic_D21_TMM')

#write_tsv(meanTMM,
#          here("results", "DEG_edgeR", 
#               "meanTMM_by_group_notmulti.tsv"))

edger.output.notmulti <- meanTMM %>% 
  full_join(chronVSctrl.D0_t, by = c("gene_id","GeneSymbol","Class")) %>% 
  full_join(chronVSctrl.D21_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronVSctrl.D0", "_chronVSctrl.D21")) %>% 
  full_join(chronD21VSchronD0_t, by = c("gene_id","GeneSymbol","Class")) %>% 
  full_join(ctrlD21VSctrlD0_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronD21VSchronD0", "_ctrlD21VSctrlD0")) %>%
  full_join(chronVSctrl_t, by = c("gene_id","GeneSymbol","Class")) %>%
  full_join(D21VSD0_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronVSctrl", "_D21VSD0"))

#write_tsv(edger.output.notmulti,
#          here("results", "DEG_edgeR", 
#               "diff_express_all_comparisons_notmulti.tsv"))

edger.output.notmulti %>% 
  filter(logFC_chronVSctrl.D0 > 1 & FDR_chronVSctrl.D0 < 0.1) %>% nrow()
edger.output.notmulti %>% 
  filter(logFC_chronVSctrl.D0 < -1 & FDR_chronVSctrl.D0 < 0.1) %>% nrow()

edger.output.notmulti %>% 
  filter(logFC_chronVSctrl.D21 > 1 & FDR_chronVSctrl.D21 < 0.1) %>% nrow()
edger.output.notmulti %>% 
  filter(logFC_chronVSctrl.D21 < -1 & FDR_chronVSctrl.D21 < 0.1) %>% nrow()

edger.output.notmulti %>% 
  filter(logFC_chronD21VSchronD0 > 1 & FDR_chronD21VSchronD0 < 0.1) %>% nrow()
edger.output.notmulti %>% 
  filter(logFC_chronD21VSchronD0 < -1 & FDR_chronD21VSchronD0 < 0.1) %>% nrow()

edger.output.notmulti %>% 
  filter(logFC_ctrlD21VSctrlD0 > 1 & FDR_ctrlD21VSctrlD0 < 0.1) %>% nrow()
edger.output.notmulti %>% 
  filter(logFC_ctrlD21VSctrlD0 < -1 & FDR_ctrlD21VSctrlD0 < 0.1) %>% nrow()

edger.output.notmulti %>% 
  filter(logFC_chronVSctrl > 1 & FDR_chronVSctrl < 0.1) %>% nrow()
edger.output.notmulti %>% 
  filter(logFC_chronVSctrl < -1 & FDR_chronVSctrl < 0.1) %>% nrow()

edger.output.notmulti %>% 
  filter(logFC_D21VSD0 > 1 & FDR_D21VSD0 < 0.1) %>% nrow()
edger.output.notmulti %>% 
  filter(logFC_D21VSD0 < -1 & FDR_D21VSD0 < 0.1) %>% nrow()

```

```{r Volcano Plot}
##### Volcano plot function (NO label)
volc_plot <- function(topTags.out, df = immgene) {
  up <- topTags.out %>% filter(logFC > 1 & FDR < 0.1)
  
  down <- topTags.out %>% filter(logFC < -1 & FDR < 0.1)
  
  topTags.out %>% 
    ggplot(aes(logFC, -log10(FDR))) +
    geom_point(color = "grey", size = 0.8) +
    geom_point(data = up ,
               aes(logFC, -log10(FDR)), color = "#EF8A62", size = 0.8) +
    geom_point(data = down,
               aes(logFC, -log10(FDR)), color = "#67A9CF", size = 0.8) +
    geom_vline(xintercept = c(-log2(2), log2(2)), linetype="dashed", 
                  color = "red", size = 0.3) +
    geom_hline(yintercept = -log10(0.1), linetype="dashed", color = "red", 
               size = 0.3) +
#    geom_text_repel(data = up %>%
#                      bind_rows(down) %>% 
#                      filter(GeneSymbol %in% df$X1), 
#                    aes(label = GeneSymbol), min.segment.length = 0,
#                    max.overlaps = Inf, size = 3) +
    theme_classic()
}

# chronVSctrl.D0
chronVSctrl.D0_vp <- volc_plot(chronVSctrl.D0_t)

# chronVSctrl.D21
chronVSctrl.D21_vp <- volc_plot(chronVSctrl.D21_t)

# chronD21VSchronD0
chronD21VSchronD0_vp <- volc_plot(chronD21VSchronD0_t)

# ctrlD21VSctrlD0
ctrlD21VSctrlD0_vp <- volc_plot(ctrlD21VSctrlD0_t)

# chronVSctrl
chronVSctrl_vp <- volc_plot(chronVSctrl_t)

# D21VSD0
D21VSD0_vp <- volc_plot(D21VSD0_t)

patchw <- (chronVSctrl.D0_vp | chronVSctrl.D21_vp) /
  (chronD21VSchronD0_vp|ctrlD21VSctrlD0_vp)/
  (chronVSctrl_vp|D21VSD0_vp)

fp <- patchw + plot_annotation(tag_levels = 'A')

ggsave(here("results","figures","volcanoPlot_all_comparisons_Notmulti.png"), fp, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2.5)

###### Volcano plot function WITH label
volc_plot_la <- function(topTags.out, df = immgene) {
  up <- topTags.out %>% filter(logFC > 1 & FDR < 0.1)
  
  down <- topTags.out %>% filter(logFC < -1 & FDR < 0.1)
  
  topTags.out %>% 
    ggplot(aes(logFC, -log10(FDR))) +
    geom_point(color = "grey") +
    geom_point(data = up ,
               aes(logFC, -log10(FDR)), color = "#EF8A62", size = 2.5) +
    geom_point(data = down,
               aes(logFC, -log10(FDR)), color = "#67A9CF", size = 2.5) +
    geom_vline(xintercept = c(-log2(2), log2(2)), linetype="dashed", 
                  color = "red", size = 0.3) +
    geom_hline(yintercept = -log10(0.1), linetype="dashed", color = "red", 
               size = 0.3) +
    geom_text_repel(data = up %>%
                      bind_rows(down) %>% 
                      filter(GeneSymbol %in% df$X1), 
                    aes(label = GeneSymbol), min.segment.length = 0,
                    max.overlaps = Inf, size = 3) +
    theme_classic()
}

# chronVSctrl.D0
chronVSctrl.D0_vp_la <- volc_plot_la(chronVSctrl.D0_t)

ggsave(here("results","figures","volcanoPlot_chronVSctrl.D0_Notmulti.png"),
       chronVSctrl.D0_vp_la, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2.5)

# chronVSctrl.D21
chronVSctrl.D21_vp_la <- volc_plot_la(chronVSctrl.D21_t)

ggsave(here("results","figures","volcanoPlot_chronVSctrl.D21_Notmulti.png"),
       chronVSctrl.D21_vp_la, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2.5)

# chronD21VSchronD0
chronD21VSchronD0_vp_la <- volc_plot_la(chronD21VSchronD0_t)

ggsave(here("results","figures","volcanoPlot_chronD21VSchronD0_Notmulti.png"),
       chronD21VSchronD0_vp_la, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2.5)

# ctrlD21VSctrlD0
ctrlD21VSctrlD0_vp_la <- volc_plot_la(ctrlD21VSctrlD0_t)

ggsave(here("results","figures","volcanoPlot_ctrlD21VSctrlD0_Notmulti.png"),
       ctrlD21VSctrlD0_vp_la, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2.5)

# chronVSctrl
chronVSctrl_vp_la <- volc_plot_la(chronVSctrl_t)

ggsave(here("results","figures","volcanoPlot_chronVSctrl_Notmulti.png"),
       chronVSctrl_vp_la, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2.5)

# D21VSD0
D21VSD0_vp_la <- volc_plot_la(D21VSD0_t)

ggsave(here("results","figures","volcanoPlot_D21VSD0_Notmulti.png"),
       D21VSD0_vp_la, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 2.5)
```

```{r Gene set enrichment analysis - Not multimapping}
## edgeR CAMERA
#term.list <- split(gene_set, as.factor(gene_set$term))
#path.list <- lapply(term.list, function(x) return(unname(as_vector(x[,2]))))
#idx <- ids2indices(path.list, id = y$genes$GeneSymbol)

idx <- ids2indices(gene_set, id = y$genes$GeneSymbol)

chronVSctrl.D0.enrich <- camera(y, idx, design, 
                                contrast = my.contrast[, "chronVSctrl.D0"]) %>% 
  rownames_to_column("pathway") %>% 
  mutate(stat = ifelse(Direction == "Up", -log10(FDR)*1, -log10(FDR)*-1))

chronVSctrl.D21.enrich <- camera(y, idx, design, 
                                 contrast = my.contrast[, "chronVSctrl.D21"]) %>% 
  rownames_to_column("pathway") %>% 
  mutate(stat = ifelse(Direction == "Up", -log10(FDR)*1, -log10(FDR)*-1))

chronD21VSchronD0.enrich <- camera(y, idx, design,
                                   contrast = my.contrast[, "chronD21VSchronD0"]) %>% 
  rownames_to_column("pathway") %>% 
  mutate(stat = ifelse(Direction == "Up", -log10(FDR)*1, -log10(FDR)*-1))

ctrlD21VSctrlD0.enrich <- camera(y, idx, design, 
                                  contrast = my.contrast[, "ctrlD21VSctrlD0"]) %>% 
  rownames_to_column("pathway") %>% 
  mutate(stat = ifelse(Direction == "Up", -log10(FDR)*1, -log10(FDR)*-1))

chronVSctrl.enrich <- camera(y, idx, design, 
                             contrast = my.contrast[, "chronVSctrl"]) %>% 
  rownames_to_column("pathway") %>% 
  mutate(stat = ifelse(Direction == "Up", -log10(FDR)*1, -log10(FDR)*-1))

D21VSD0.enrich <- camera(y, idx, design,
                             contrast = my.contrast[, "D21VSD0"]) %>% 
  rownames_to_column("pathway") %>% 
  mutate(stat = ifelse(Direction == "Up", -log10(FDR)*1, -log10(FDR)*-1))


barcodeplot(D21VSD0_t$logFC,
            index=idx[["G2/M Checkpoints"]],
            labels=c("D21","D0"),
            main="G2/M Checkpoints",
            alpha=1)

stat <- chronVSctrl.D0.enrich %>% 
  slice_head(n = 10) %>% 
  dplyr::select(pathway, stat) %>% 
  full_join(chronVSctrl.D21.enrich %>%
              slice_head(n = 10) %>% 
               dplyr::select(pathway, stat), by = "pathway", 
            suffix = c("_chronVSctrl.D0", "_chronVSctrl.D21")) %>% 
  full_join(chronVSctrl.enrich %>%
              slice_head(n = 10) %>%
              dplyr::select(pathway, stat), by = "pathway") %>% 
  full_join(chronD21VSchronD0.enrich %>% 
              slice_head(n = 10) %>%
               dplyr::select(pathway, stat), by = "pathway", 
            suffix = c("_chronVSctrl", "_chronD21VSchronD0")) %>%
  full_join(ctrlD21VSctrlD0.enrich %>% 
              slice_head(n = 10) %>%
               dplyr::select(pathway, stat), by = "pathway") %>%
  full_join(D21VSD0.enrich %>% 
              slice_head(n = 10) %>%
               dplyr::select(pathway, stat), by = "pathway",
            suffix = c("_ctrlD21VSctrlD0", "_D21VSD0")) %>% 
  column_to_rownames("pathway") %>% 
  rename_all(~ sub("stat_", "", .x)) %>% 
  mutate_all(~ case_when(is.na(.) ~ 0, TRUE ~ .))
  

## fgsea

# chronVSctrl.D0
ranks_chronVSctrl.D0 <- chronVSctrl.D0_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chronVSctrl.D0 <- fgsea(pathways=gene_set, 
                                       stats=ranks_chronVSctrl.D0, eps=0)

# chronVSctrl.D21
ranks_chronVSctrl.D21 <- chronVSctrl.D21_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chronVSctrl.D21 <- fgsea(pathways=gene_set, 
                                       stats=ranks_chronVSctrl.D21, eps=0)

# chronVSctrl
ranks_chronVSctrl <- chronVSctrl_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chronVSctrl <- fgsea(pathways=gene_set, 
                                       stats=ranks_chronVSctrl, eps=0)

# chronD21VSchronD0
ranks_chronD21VSchronD0 <- chronD21VSchronD0_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_chronD21VSchronD0 <- fgsea(pathways=gene_set, 
                                    stats=ranks_chronD21VSchronD0, eps=0)

# ctrlD21VSctrlD0
ranks_ctrlD21VSctrlD0 <- ctrlD21VSctrlD0_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_ctrlD21VSctrlD0 <- fgsea(pathways=gene_set, 
                                    stats=ranks_ctrlD21VSctrlD0, eps=0)

# D21VSD0
ranks_D21VSD0 <- D21VSD0_t %>% 
  mutate(rank = -log10(PValue)/sign(logFC)) %>% 
  dplyr::select(GeneSymbol, rank) %>% 
  group_by(GeneSymbol) %>% 
  summarize(rank=mean(rank)) %>% 
  deframe()

fgseaRes_D21VSD0 <- fgsea(pathways=gene_set, stats=ranks_D21VSD0, eps=0)

# Join fgsea results
fgseaRes <- fgseaRes_chronVSctrl.D0 %>% 
  as_tibble() %>% 
  select(pathway, padj, NES) %>% 
  full_join(fgseaRes_chronVSctrl.D21 %>% 
              as_tibble() %>% 
              select(pathway, padj, NES), by = "pathway", 
            suffix = c("_chronVSctrl.D0", "_chronVSctrl.D21")) %>% 
  full_join(fgseaRes_chronVSctrl %>% 
              as_tibble() %>% 
              select(pathway, padj, NES), by = "pathway") %>% 
  full_join(fgseaRes_chronD21VSchronD0 %>% 
              as_tibble() %>% 
              select(pathway, padj, NES), by = "pathway", 
            suffix = c("_chronVSctrl", "_chronD21VSchronD0")) %>%
  full_join(fgseaRes_ctrlD21VSctrlD0 %>% 
              as_tibble() %>% 
              select(pathway, padj, NES), by = "pathway") %>%
  full_join(fgseaRes_D21VSD0 %>% 
              as_tibble() %>% 
              select(pathway, padj, NES), by = "pathway",
            suffix = c("_ctrlD21VSctrlD0", "_D21VSD0"))

nes <- fgseaRes %>% 
  column_to_rownames("pathway") %>% 
  dplyr::select(starts_with("NES"))

padj <- fgseaRes %>% 
  column_to_rownames("pathway") %>% 
  dplyr::select(starts_with("padj"))

#fgseaRes_Tidy_chronVSctrl <- fgseaRes_chronVSctrl.D0 %>% 
#  as_tibble() %>% 
#  select(pathway, padj, NES) %>% 
#  full_join(fgseaRes_chronVSctrl.D21 %>% 
#              as_tibble() %>% 
#              select(pathway, padj, NES), by = "pathway", 
#            suffix = c("_chronVSctrl.D0", "_chronVSctrl.D21")) %>% 
#  full_join(fgseaRes_chronVSctrl %>% 
#              as_tibble() %>% 
#              select(pathway, padj, NES), by = "pathway") %>% 
#  dplyr::rename(padj_chronVSctrl = padj, NES_chronVSctrl = NES)

#fgseaRes_Tidy_D21VSD0 <- fgseaRes_chronD21VSchronD0 %>% 
#  as_tibble() %>% 
#  select(pathway, padj, NES) %>%  
#  full_join(fgseaRes_ctrlD21VSctrlD0 %>% 
#              as_tibble() %>% 
#              select(pathway, padj, NES), by = "pathway",
#            suffix = c("_chronD21VSchronD0", "_ctrlD21VSctrlD0")) %>%
#  full_join(fgseaRes_D21VSD0 %>% 
#              as_tibble() %>% 
#              select(pathway, padj, NES), by = "pathway") %>% 
#  dplyr::rename(padj_D21VSD0 = padj, NES_D21VSD0 = NES)
  
#nes_chronVSctrl <- fgseaRes_Tidy_chronVSctrl %>% 
#  column_to_rownames("pathway") %>% 
#  dplyr::select(starts_with("NES"))

#padj_chronVSctrl <- fgseaRes_Tidy_chronVSctrl %>% 
#  column_to_rownames("pathway") %>% 
#  dplyr::select(starts_with("padj"))

#ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
#  geom_col(aes(fill=padj<0.05)) +
#  coord_flip() +
#  labs(x="Pathway", y="Normalized Enrichment Score",
#       title="Hallmark pathways NES from GSEA") + 
#  theme_minimal()

#plotEnrichment(pathways.hallmark[[head(fgseaRes[order(pval), ], 1)$pathway]],
#               ranks) + labs(title=head(fgseaRes[order(pval), ], 1)$pathway)
```

```{r GSEA Heatmap}
pcut = 0.05

nes <- stat

nes <- nes[rowSums(nes > -log10(pcut) | nes < -log10(pcut) *-1) >= 1, , drop=FALSE]

custom_pal <- c("#053061", "#2166AC", "#4393C3", "#92C5DE",
                "#D1E5F0", "#FFFFFF", "#FDDBC7", "#F4A582",
                "#D6604D", "#B2182B", "#67001F")

custom_pal <- colorRampPalette(custom_pal)(200)

nes <- as.matrix(nes)

nes[which(nes < -log10(pcut) & nes > -log10(pcut)*-1, arr.ind=TRUE)] <- 0

#if(nrow(nes) > 2){
#  row_order <- rownames(nes)[hclust(dist(nes))$order]
#} else {
#  row_order <- rownames(nes)
#}

nes_melted <- reshape2::melt(nes)
colnames(nes_melted) <- c("Pathway", "Class", "stat")

nes_melted$Pathway <- factor(nes_melted$Pathway)
nes_melted$Class <- factor(nes_melted$Class, 
                           levels = c("chronVSctrl.D0", "chronVSctrl.D21", "chronVSctrl",
                                      "chronD21VSchronD0", "ctrlD21VSctrlD0", "D21VSD0"))

max_abs_nes <- max(abs(nes_melted$stat))
        


res <- ggplot(nes_melted, aes_(x=~Class, y=~Pathway, size=~abs(stat), fill=~stat)) +
  geom_point(color = "white", shape=21) +
  scale_fill_gradientn(colours=custom_pal, space = "Lab",
                       limits=c(-max_abs_nes, max_abs_nes)) +
  scale_size(range=c(0,5), limits=c(0, NA)) +
  guides(size="none") +
  theme_minimal() +
  theme(panel.grid.major = element_blank(), 
        axis.text.x = element_text(angle = 30, vjust = 1, hjust=1)) 


ggsave(here("results","figures","enrichPlot_DEGS_NotMulti.png"), res, 
       device = "png", width = 9, height = 6.5, units = "cm", 
       dpi = "print", scale = 2)


############## FGSEA

pcut = 0.0001

padj <- padj[rowSums(padj < pcut) >= 1, , drop=FALSE]
nes <- nes[rownames(padj), , drop=FALSE]

custom_pal <- c("#053061", "#2166AC", "#4393C3", "#92C5DE",
                "#D1E5F0", "#FFFFFF", "#FDDBC7", "#F4A582",
                "#D6604D", "#B2182B", "#67001F")

custom_pal <- colorRampPalette(custom_pal)(200)

nes <- as.matrix(nes)
padj <- as.matrix(padj)
nes[which(padj > pcut, arr.ind=TRUE)] <- 0

if(nrow(nes) > 2){
  row_order <- rownames(nes)[hclust(dist(nes))$order]
} else {
  row_order <- rownames(nes)
}

nes_melted <- reshape2::melt(nes)
colnames(nes_melted) <- c("Module", "Class", "NES")
nes_melted$Module <- factor(nes_melted$Module, levels=row_order)
nes_melted$Class <- as.character(nes_melted$Class)
max_abs_nes <- max(abs(nes_melted$NES))
        
ggplot(nes_melted, aes_(x=~Class, y=~Module, size=~abs(NES), fill=~NES)) +
  geom_point(color = "white", shape=21) +
  scale_fill_gradientn(colours=custom_pal, space = "Lab",
                       limits=c(-max_abs_nes, max_abs_nes)) +
  scale_size(range=c(0,5), limits=c(0, NA)) +
  guides(size="none") +
  theme_minimal() +
  theme(panel.grid.major = element_blank()) 

```

```{r DE analysis Multimapping}
group <- factor(paste(pheno$Class, pheno$Day, sep = "_"))

#Not multimapping
y <- DGEList(counts = rawcount_multi[,-1], 
             genes = hannot[,c(1,2,4)],
             group = group)

design <- model.matrix(~group+0)

# Filtering low expression genes
keep <- filterByExpr(y, min.count = 5)
y <- y[keep, , keep.lib.sizes=FALSE]

# Norm factors
y <- calcNormFactors(y)
#plotMDS(y, col = brewer.pal(4,"Set2")[group], labels = group)

# Estimate Dispersion
y <- estimateDisp(y, design)
#plotBCV(y)

#DE
fit <- glmQLFit(y, design)

chronVSctrl.D0 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSctrl.D0"])
chronVSctrl.D21 <- glmQLFTest(fit, contrast = my.contrast[, "chronVSctrl.D21"])
chronD21VSchronD0 <- glmQLFTest(fit, contrast = my.contrast[, "chronD21VSchronD0"])
ctrlD21VSctrlD0 <- glmQLFTest(fit, contrast = my.contrast[, "ctrlD21VSctrlD0"])
chronVSctrl <- glmQLFTest(fit, contrast = my.contrast[, "chronVSctrl"])
D21VSD0 <- glmQLFTest(fit, contrast = my.contrast[, "D21VSD0"])

chronVSctrl.D0_t <- topTags(chronVSctrl.D0, n = NULL)$table
chronVSctrl.D21_t <- topTags(chronVSctrl.D21, n = NULL)$table
chronD21VSchronD0_t <- topTags(chronD21VSchronD0, n = NULL)$table
ctrlD21VSctrlD0_t <- topTags(ctrlD21VSctrlD0, n = NULL)$table
chronVSctrl_t <- topTags(chronVSctrl, n = NULL)$table
D21VSD0_t <- topTags(D21VSD0, n = NULL)$table

edger.output.multi <- chronVSctrl.D0_t %>%
  full_join(chronVSctrl.D21_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronVSctrl.D0", "_chronVSctrl.D21")) %>% 
  full_join(chronD21VSchronD0_t, by = c("gene_id","GeneSymbol","Class")) %>% 
  full_join(ctrlD21VSctrlD0_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronD21VSchronD0", "_ctrlD21VSctrlD0")) %>%
  full_join(chronVSctrl_t, by = c("gene_id","GeneSymbol","Class")) %>%
  full_join(D21VSD0_t, by = c("gene_id","GeneSymbol","Class"), 
            suffix = c("_chronVSctrl", "_D21VSD0"))

#write_tsv(edger.output.multi,
#          here("results", "DEG_edgeR", 
#               "diff_express_all_comparisons_multi.tsv"))

edger.output.multi %>% 
  filter(logFC_chronVSctrl.D0 > 1 & FDR_chronVSctrl.D0 < 0.1) %>% nrow()
edger.output.multi %>% 
  filter(logFC_chronVSctrl.D0 < -1 & FDR_chronVSctrl.D0 < 0.1) %>% nrow()

edger.output.multi %>% 
  filter(logFC_chronVSctrl.D21 > 1 & FDR_chronVSctrl.D21 < 0.1) %>% nrow()
edger.output.multi %>% 
  filter(logFC_chronVSctrl.D21 < -1 & FDR_chronVSctrl.D21 < 0.1) %>% nrow()

edger.output.multi %>% 
  filter(logFC_chronD21VSchronD0 > 1 & FDR_chronD21VSchronD0 < 0.1) %>% nrow()
edger.output.multi %>% 
  filter(logFC_chronD21VSchronD0 < -1 & FDR_chronD21VSchronD0 < 0.1) %>% nrow()

edger.output.multi %>% 
  filter(logFC_ctrlD21VSctrlD0 > 1 & FDR_ctrlD21VSctrlD0 < 0.1) %>% nrow()
edger.output.multi %>% 
  filter(logFC_ctrlD21VSctrlD0 < -1 & FDR_ctrlD21VSctrlD0 < 0.1) %>% nrow()

edger.output.multi %>% 
  filter(logFC_chronVSctrl > 1 & FDR_chronVSctrl < 0.1) %>% nrow()
edger.output.multi %>% 
  filter(logFC_chronVSctrl < -1 & FDR_chronVSctrl < 0.1) %>% nrow()

edger.output.multi %>% 
  filter(logFC_D21VSD0 > 1 & FDR_D21VSD0 < 0.1) %>% nrow()
edger.output.multi %>% 
  filter(logFC_D21VSD0 < -1 & FDR_D21VSD0 < 0.1) %>% nrow()
```

```{r Plot individual genes for inspection}
x <- outTPMnotmulti %>% 
  pivot_longer(-c('Symbol','GeneSymbol'), names_to = 'lib', 
               values_to = 'tpm') %>%
  mutate(lib = as.double(lib)) %>% 
  full_join(pheno, by = c('lib' = 'Sample')) %>% 
  unite("ClassDay", Class:Day)

x %>% 
  filter(GeneSymbol == "EIF1AY") %>% 
  ggplot(aes(ClassDay, log2(tpm+1))) +
  geom_boxplot()
```

```{r Interferon/Interleukine plot}
tidytpm_notmulti <- outTPMnotmulti %>%
  filter(GeneSymbol %in% ilifn$geneSymbol) %>% 
  pivot_longer(-c('Symbol','GeneSymbol'), names_to = 'lib', 
               values_to = 'tpm') %>%
  mutate(lib = as.double(lib)) %>% 
  full_join(pheno, by = c('lib' = 'Sample')) %>% 
  unite("ClassDay", Class:Day) %>% 
  mutate(protocol = "NotMulti")

tidytpm_multi <- outTPMmulti %>%
  filter(GeneSymbol %in% ilifn$geneSymbol) %>% 
  pivot_longer(-c('Symbol','GeneSymbol'), names_to = 'lib', 
               values_to = 'tpm') %>%
  mutate(lib = as.double(lib)) %>% 
  full_join(pheno, by = c('lib' = 'Sample')) %>% 
  unite("ClassDay", Class:Day) %>% 
  mutate(protocol = "Multi")

tidytpm <- tidytpm_multi %>%
  bind_rows(tidytpm_notmulti)


# Violin Plot 75 genes  
tidytpm %>%
  ggplot(aes(ClassDay, log10(tpm+1), fill = protocol)) +
  geom_violin() +
  facet_wrap(~GeneSymbol) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

# Correlation plot
corplot <- tidytpm  %>% 
  pivot_wider(names_from = protocol, values_from = tpm) %>% 
  ggplot(aes(x = log2(Multi + 1), y = log2(NotMulti +1), )) +
  geom_point() +
  stat_cor(label.y = 9.5, digits = 3) +
  geom_smooth(method='lm', formula= y~x) +
  theme_bw()

ggsave(here("results","figures","corplot_Multi_Notmulti.png"), corplot, 
       device = "png", width = 10, height = 6, units = "cm", 
       dpi = "print", scale = 1.5)

```

